### DISCUSSION TITLE
What can we learn from Lean?

### DISCUSSION THREAD

**Discussion by nmvdw** [UniMath/UniMath#1778] [Ideas]
I talked with a Lean maintainer today (Johan Commelin), and he showed me how Lean manages pull requests. They have several 'linters'. These automatically check whether the files are properly documented (they have guidelines for it), they check for a couple of style aspects, they check whether the lines are too long, and so on. In addition, they have a system using labels for pull requests: 'review awaiting', 'response awaiting' (I forgot the precise names), so that people know what they can expect from reviewers. They also have a system to keep PRs small: PRs can be split into multiple PRs with a dependency. This eases the job for the code reviewers.

If there are more contributors to UniMath, then such systems are essential. It helps the reviewers to make the job doable, and it helps the contributors by making it easy for them to get quick feedback to their PRs.

As such, I think there's a great opportunity for us to learn from Lean, and I think that we should look at how they deal with certain problems, learn from them, so that we can also use this to improve UniMath. 

What do you all think about this? Note: I dont have any details on how to implement this precisely, but I think it is a good thing to look into.

**Comment by m-lindgren** [comment]
These are some (hastily written down) ideas related to this that I would like to implement and test in October/November:

1. Change the review process order to only do the time consuming builds once a PR is scheduled to be merged (with the possibility of triggering builds manually). This can probably be done with githubs merge queue function and some clever use of labels and CI scripts.
2. One exception to the above is that satellites are always rebuilt, this is because many users can not test build satellites locally.
3. Have build-dependencies across repositories automatically tested against each other. Say PR UniMath#9999 to UniMath breaks TypeTheory, but the user also submits a PR TypeTheory#90 to TypeTheory, then the user could add a tag/label/info field in both PR:s that references each other. The CI would then build UniMath#9999 not against TypeTheory/master but against TypeTheory with TypeTheory#90 applied (and vice versa).
4. Have build-dependencies across repositories automatically merged if they all are approved and the PR to UniMath gets merged. So in the example above, if UniMath#9999 is merged then TypeTheory#90 will automatically be merged into TypeTheory once CI is done with compiling and merging UniMath#9999.
5. Allow PR dependencies, with optional automatic merging. Similar to the two points above but internal to UniMath only. This should allow a  user who submits a PR that gets blocked during review for some reason to continue development and submit PR:s that depends on the first. This would allow us to review/test/approve/suggest changes to PR:s that can not be merged yet because a blocked dependency. The optional automatic merging can be done with a label saying that if the dependencies are merged then this can also be merged. With this we could also encourage smaller PR:s to be submitted, and large PR:s to be split. Key point why this would be useful: You don't have to wait for your first PR to be merged before you submit your follow-up PR. Your follow-up PR will be built and tested against your PR dependency, not against master.
6. And just so  I don't forget, but users should not have to deal with CONTENTS and .package/files stuff.

Assuming this is possible to implement I suspect this should _significantly_ reduce the number of rebuilds made and cut down on some of the CO2 production. I also hope this would provide some quality of life improvements for contributors of UniMath.

**Comment by nmvdw** [comment]
I like all the points that you mentioned.

**Comment by nmvdw** [comment]
Starting in October, I would like to put in some effort in writing documentation for the 'Bicategories' directory, and develop documentation standards for UniMath.
